### PR TITLE
[Shared] Remove isRequired property serialization

### PR DESCRIPTION
**Pull request by almedina-ms**
Fixes https://github.com/microsoft/AdaptiveCards/issues/3072

Removes the serialization of isRequired so the android tests work correctly but doesn't remove the getter and setter methods to avoid any breaking changes, also, as Input Validation is coming to v1.3 those methods will be removed in the future

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/AdaptiveCards/pull/3131)

### CODE REVIEW

**Comment by paulcam206** [comment]
looks like tests are failing:

```
Failed   EverythingBagelTests
Error Message:
 Assert failed. Expected:<{"actions":[{"data":{"submitValue":true},"id":"Action.Submit_id","title":"Action.Submit","type":"Action.Submit"},{"card":{"actions":[],"backgroundImage":{"fillMode":"repeat","horizontalAlignment":"right","url":"https://adaptivecards.io/content/cats/1.png","verticalAlignment":"center"},"body":[{"isSubtle":true,"text":"Action.ShowCard text","type":"TextBlock"}],"lang":"en","type":"AdaptiveCard","version":"1.0"},"id":"Action.ShowCard_id","title":"Action.ShowCard","type":"Action.ShowCard"}],"backgroundImage":"https://adaptivecards.io/content/cats/1.png","body":[{"id":"TextBlock_id","italic":true,"maxLines":1,"strikethrough":true,"text":"TextBlock_text","type":"TextBlock"},{"fontType":"Monospace","id":"TextBlock_id_mono","italic":true,"maxLines":1,"strikethrough":true,"text":"TextBlock_text","type":"TextBlock"},{"id":"TextBlock_id_def","italic":true,"maxLines":1,"strikethrough":true,"text":"TextBlock_text","type":"TextBlock"},{"altText":"Image_altText","horizontalAlignment":"center","id":"Image_id","isVisible":false,"selectAction":{"title":"Image_Action.OpenUrl","type":"Action.OpenUrl","url":"https://adaptivecards.io/"},"separator":true,"size":"Auto","spacing":"none","style":"person","type":"Image","url":"https://adaptivecards.io/content/cats/1.png"},{"id":"Container_id","items":[{"columns":[{"id":"Column_id1","items":[{"type":"Image","url":"https://adaptivecards.io/content/cats/1.png"}],"style":"Default","type":"Column","width":"auto"},{"id":"Column_id2","items":[{"type":"Image","url":"https://adaptivecards.io/content/cats/2.png"}],"style":"Emphasis","type":"Column","width":"20px"},{"id":"Column_id3","items":[{"type":"Image","url":"https://adaptivecards.io/content/cats/3.png"},{"id":"Column3_TextBlock_id","text":"Column3_TextBlock_text","type":"TextBlock"}],"style":"Default","type":"Column","width":"stretch"}],"id":"ColumnSet_id","separator":true,"spacing":"large","type":"ColumnSet"}],"selectAction":{"data":"Container_data","title":"Container_Action.Submit","type":"Action.Submit"},"spacing":
Stack Trace:
	at AdaptiveCardsSharedModelUnitTest::EverythingBagel::EverythingBagelTests() in c:\ba\848\s\source\shared\cpp\adaptivecardssharedmodel\adaptivecardssharedmodelunittest\everythingbagel.cpp:line 556
Standard Output Messages:
 Expected: '{"actions":[{"data":{"submitValue":true},"id":"Action.Submit_id","title":"Action.Submit","type":"Action.Submit"},{"card":{"actions":[],"backgroundImage":{"fillMode":"repeat","horizontalAlignment":"right","url":"https://adaptivecards.io/content/cats/1.png","verticalAlignment":"center"},"body":[{"isSubtle":true,"text":"Action.ShowCard text","type":"TextBlock"}],"lang":"en","type":"AdaptiveCard","version":"1.0"},"id":"Action.ShowCard_id","title":"Action.ShowCard","type":"Action.ShowCard"}],"backgroundImage":"https://adaptivecards.io/content/cats/1.png","body":[{"id":"TextBlock_id","italic":true,"maxLines":1,"strikethrough":true,"text":"TextBlock_text","type":"TextBlock"},{"fontType":"Monospace","id":"TextBlock_id_mono","italic":true,"maxLines":1,"strikethrough":true,"text":"TextBlock_text","type":"TextBlock"},{"id":"TextBlock_id_def","italic":true,"maxLines":1,"strikethrough":true,"text":"TextBlock_text","type":"TextBlock"},{"altText":"Image_altText","horizontalAlignment":"center","id":"Image_id","isVisible":false,"selectAction":{"title":"Image_Action.OpenUrl","type":"Action.OpenUrl","url":"https://adaptivecards.io/"},"separator":true,"size":"Auto","spacing":"none","style":"person","type":"Image","url":"https://adaptivecards.io/content/cats/1.png"},{"id":"Container_id","items":[{"columns":[{"id":"Column_id1","items":[{"type":"Image","url":"https://adaptivecards.io/content/cats/1.png"}],"style":"Default","type":"Column","width":"auto"},{"id":"Column_id2","items":[{"type":"Image","url":"https://adaptivecards.io/content/cats/2.png"}],"style":"Emphasis","type":"Column","width":"20px"},{"id":"Column_id3","items":[{"type":"Image","url":"https://adaptivecards.io/content/cats/3.png"},{"id":"Column3_TextBlock_id","text":"Column3_TextBlock_text","type":"TextBlock"}],"style":"Default","type":"Column","width":"stretch"}],"id":"ColumnSet_id","separator":true,"spacing":"large","type":"ColumnSet"}],"selectAction":{"data":"Container_data","title":"Container_Action.Submit","type":"Action.Submit"},"spacing":"medium","style":"Default","type":"Container"},{"facts":[{"title":"Topping","value":"poppyseeds"},{"title":"Topping","value":"onion flakes"}],"id":"FactSet_id","type":"FactSet"},{"id":"ImageSet_id","imageSize":"Auto","images":[{"type":"Image","url":"https://adaptivecards.io/content/cats/1.png"},{"type":"Image","url":"https://adaptivecards.io/content/cats/2.png"},{"type":"Image","url":"https://adaptivecards.io/content/cats/3.png"}],"separator":true,"type":"ImageSet"},{"id":"Container_id_inputs","items":[{"id":"Input.Text_id","inlineAction":{"iconUrl":"https://adaptivecards.io/content/cats/1.png","title":"Input.Text_Action.Submit","type":"Action.Submit"},"isRequired":false,"maxLength":10,"placeholder":"Input.Text_placeholder","spacing":"small","style":"text","type":"Input.Text","value":"Input.Text_value"},{"id":"Input.Number_id","isRequired":false,"max":10,"min":5,"placeholder":"Input.Number_placeholder","type":"Input.Number","value":7},{"id":"Input.Date_id","isRequired":false,"max":"1/1/2020","min":"8/1/2018","placeholder":"Input.Date_placeholder","type":"Input.Date","value":"8/9/2018"},{"id":"Input.Time_id","isRequired":false,"max":"17:00","min":"10:00","placeholder":"Input.Time_placeholder","type":"Input.Time","value":"13:00"},{"id":"Input.Toggle_id","isRequired":false,"title":"Input.Toggle_title","type":"Input.Toggle","value":"Input.Toggle_on","valueOff":"Input.Toggle_off","valueOn":"Input.Toggle_on"},{"size":"Large","text":"Everybody's got choices","type":"TextBlock","weight":"Bolder"},{"choices":[{"title":"Input.Choice1_title","value":"Input.Choice1"},{"title":"Input.Choice2_title","value":"Input.Choice2"},{"title":"Input.Choice3_title","value":"Input.Choice3"},{"title":"Input.Choice4_title","value":"Input.Choice4"}],"id":"Input.ChoiceSet_id","isMultiSelect":true,"isRequired":false,"style":"Compact","type":"Input.ChoiceSet","value":"Input.Choice2,Input.Choice4"}],"type":"Container"},{"actions":[{"id":"ActionSet.Action.Submit_id","title":"ActionSet.Action.Submit","type":"Action.Submit"},{"id":"ActionSet.Action.OpenUrl_id","title":"ActionSet.Action.OpenUrl","type":"Action.OpenUrl","url":"https://adaptivecards.io/"}],"type":"ActionSet"},{"horizontalAlignment":"right","id":"RichTextBlock_id","inlines":[{"color":"Dark","fontType":"Monospace","highlight":true,"isSubtle":true,"italic":true,"size":"Large","strikethrough":true,"text":"This is a text run","type":"TextRun","weight":"Bolder"},{"selectAction":{"type":"Action.Submit"},"text":"This is another text run","type":"TextRun"},{"text":"This is a text run specified as a string","type":"TextRun"}],"type":"RichTextBlock"}],"fallbackText":"fallbackText","lang":"en","speak":"speak","type":"AdaptiveCard","version":"1.0"}
' Actual: '{"actions":[{"data":{"submitValue":true},"id":"Action.Submit_id","title":"Action.Submit","type":"Action.Submit"},{"card":{"actions":[],"backgroundImage":{"fillMode":"repeat","horizontalAlignment":"right","url":"https://adaptivecards.io/content/cats/1.png","verticalAlignment":"center"},"body":[{"isSubtle":true,"text":"Action.ShowCard text","type":"TextBlock"}],"lang":"en","type":"AdaptiveCard","version":"1.0"},"id":"Action.ShowCard_id","title":"Action.ShowCard","type":"Action.ShowCard"}],"backgroundImage":"https://adaptivecards.io/content/cats/1.png","body":[{"id":"TextBlock_id","italic":true,"maxLines":1,"strikethrough":true,"text":"TextBlock_text","type":"TextBlock"},{"fontType":"Monospace","id":"TextBlock_id_mono","italic":true,"maxLines":1,"strikethrough":true,"text":"TextBlock_text","type":"TextBlock"},{"id":"TextBlock_id_def","italic":true,"maxLines":1,"strikethrough":true,"text":"TextBlock_text","type":"TextBlock"},{"altText":"Image_altText","horizontalAlignment":"center","id":"Image_id","isVisible":false,"selectAction":{"title":"Image_Action.OpenUrl","type":"Action.OpenUrl","url":"https://adaptivecards.io/"},"separator":true,"size":"Auto","spacing":"none","style":"person","type":"Image","url":"https://adaptivecards.io/content/cats/1.png"},{"id":"Container_id","items":[{"columns":[{"id":"Column_id1","items":[{"type":"Image","url":"https://adaptivecards.io/content/cats/1.png"}],"style":"Default","type":"Column","width":"auto"},{"id":"Column_id2","items":[{"type":"Image","url":"https://adaptivecards.io/content/cats/2.png"}],"style":"Emphasis","type":"Column","width":"20px"},{"id":"Column_id3","items":[{"type":"Image","url":"https://adaptivecards.io/content/cats/3.png"},{"id":"Column3_TextBlock_id","text":"Column3_TextBlock_text","type":"TextBlock"}],"style":"Default","type":"Column","width":"stretch"}],"id":"ColumnSet_id","separator":true,"spacing":"large","type":"ColumnSet"}],"selectAction":{"data":"Container_data","title":"Container_Action.Submit","type":"Action.Submit"},"spacing":"medium","style":"Default","type":"Container"},{"facts":[{"title":"Topping","value":"poppyseeds"},{"title":"Topping","value":"onion flakes"}],"id":"FactSet_id","type":"FactSet"},{"id":"ImageSet_id","imageSize":"Auto","images":[{"type":"Image","url":"https://adaptivecards.io/content/cats/1.png"},{"type":"Image","url":"https://adaptivecards.io/content/cats/2.png"},{"type":"Image","url":"https://adaptivecards.io/content/cats/3.png"}],"separator":true,"type":"ImageSet"},{"id":"Container_id_inputs","items":[{"id":"Input.Text_id","inlineAction":{"iconUrl":"https://adaptivecards.io/content/cats/1.png","title":"Input.Text_Action.Submit","type":"Action.Submit"},"maxLength":10,"placeholder":"Input.Text_placeholder","spacing":"small","style":"text","type":"Input.Text","value":"Input.Text_value"},{"id":"Input.Number_id","max":10,"min":5,"placeholder":"Input.Number_placeholder","type":"Input.Number","value":7},{"id":"Input.Date_id","max":"1/1/2020","min":"8/1/2018","placeholder":"Input.Date_placeholder","type":"Input.Date","value":"8/9/2018"},{"id":"Input.Time_id","max":"17:00","min":"10:00","placeholder":"Input.Time_placeholder","type":"Input.Time","value":"13:00"},{"id":"Input.Toggle_id","title":"Input.Toggle_title","type":"Input.Toggle","value":"Input.Toggle_on","valueOff":"Input.Toggle_off","valueOn":"Input.Toggle_on"},{"size":"Large","text":"Everybody's got choices","type":"TextBlock","weight":"Bolder"},{"choices":[{"title":"Input.Choice1_title","value":"Input.Choice1"},{"title":"Input.Choice2_title","value":"Input.Choice2"},{"title":"Input.Choice3_title","value":"Input.Choice3"},{"title":"Input.Choice4_title","value":"Input.Choice4"}],"id":"Input.ChoiceSet_id","isMultiSelect":true,"style":"Compact","type":"Input.ChoiceSet","value":"Input.Choice2,Input.Choice4"}],"type":"Container"},{"actions":[{"id":"ActionSet.Action.Submit_id","title":"ActionSet.Action.Submit","type":"Action.Submit"},{"id":"ActionSet.Action.OpenUrl_id","title":"ActionSet.Action.OpenUrl","type":"Action.OpenUrl","url":"https://adaptivecards.io/"}],"type":"ActionSet"},{"horizontalAlignment":"right","id":"RichTextBlock_id","inlines":[{"color":"Dark","fontType":"Monospace","highlight":true,"isSubtle":true,"italic":true,"size":"Large","strikethrough":true,"text":"This is a text run","type":"TextRun","weight":"Bolder"},{"selectAction":{"type":"Action.Submit"},"text":"This is another text run","type":"TextRun"},{"text":"This is a text run specified as a string","type":"TextRun"}],"type":"RichTextBlock"}],"fallbackText":"fallbackText","lang":"en","speak":"speak","type":"AdaptiveCard","version":"1.0"}
'
```

File: source/shared/cpp/ObjectModel/BaseInputElement.cpp [hunk 1]
```diff
@@ -29,7 +29,5 @@ Json::Value BaseInputElement::SerializeToJsonValue() const
 {
     Json::Value root = BaseCardElement::SerializeToJsonValue();
 
-    root[AdaptiveCardSchemaKeyToString(AdaptiveCardSchemaKey::IsRequired)] = m_isRequired;
```
Comment (RebeccaAnne) [inline source/shared/cpp/ObjectModel/BaseInputElement.cpp]: So, the latest thinking on this, as of, i think yesterday, is that this property will be part of the input validation story. So given that, i'm inclined not to remove it. We've shipped it with every release, and we're going to actually hook it up to something for 1.3, so i don't think there's any reason to pull it (although it should not serialize if it's false).

Comment (RebeccaAnne) [reply source/shared/cpp/ObjectModel/BaseInputElement.cpp]: [https://github.com/microsoft/AdaptiveCards/issues/3081](https://github.com/microsoft/AdaptiveCards/issues/3081)

---
In reply to: [300051174](https://github.com/microsoft/AdaptiveCards/pull/3131#discussion_r300051174) [](ancestors = 300051174)

Comment (shalinijoshi19) [reply source/shared/cpp/ObjectModel/BaseInputElement.cpp]: @RebeccaAnne / @almedina-ms  does this mean we can abandon this PR/ link and close out #3072 ? If so, then let's do that!

Comment (almedina-ms) [reply source/shared/cpp/ObjectModel/BaseInputElement.cpp]: I think the better answer should be to add a validation for 

if(!m_isRequired){
Do serialization
}

is that right, @RebeccaAnne ??

Comment (RebeccaAnne) [reply source/shared/cpp/ObjectModel/BaseInputElement.cpp]: Yep, that is the correct behavior, and is something that needs to happen either for this bug or as part of the Input Validation work. 

---
In reply to: [300798123](https://github.com/microsoft/AdaptiveCards/pull/3131#discussion_r300798123) [](ancestors = 300798123)

File: source/shared/cpp/ObjectModel/BaseInputElement.cpp [hunk 2]
```diff
@@ -29,7 +29,10 @@ Json::Value BaseInputElement::SerializeToJsonValue() const
...
+    if (m_isRequired)
+    {
+        root[AdaptiveCardSchemaKeyToString(AdaptiveCardSchemaKey::IsRequired)] = m_isRequired;
+    }
```
Comment (RebeccaAnne) [inline source/shared/cpp/ObjectModel/BaseInputElement.cpp]: This change will require updates to the UWP test app and unit tests to update the expected behavior.

Comment (RebeccaAnne) [reply source/shared/cpp/ObjectModel/BaseInputElement.cpp]: Also, btw, i have this locally as part of my InputValidation "prototyping" so if we don't need this for 1.2.1 we can drop this change and wait for that to come along.

---
In reply to: [300811694](https://github.com/microsoft/AdaptiveCards/pull/3131#discussion_r300811694) [](ancestors = 300811694)

**Comment by RebeccaAnne** [comment]
UWP unit tests need to be updated as well. Look at source\uwp\UWPUnitTests\ObjectModelTests.cs

**Pull request merged**